### PR TITLE
Add optional caching

### PR DESCRIPTION
**Pull request by GeorgeR227**
This adds a keyword `can_prealloc` that when set to false turns off preallocations and only uses out of place versions of functions.

### CODE REVIEW

File: test/simulation.jl [hunk 1]
```diff
@@ -133,6 +133,22 @@ fl_res = copy(du.C)
...
+
+# Test same but with no preallocating
+
+f = evalsim(DiffusionWithLiteral, can_prealloc=false)
+f_noalloc = f(torus, generate)
+
+f = evalsim(DiffusionWithLiteral)
+f_alloc = f(torus, generate)
+
+f_noalloc(du, u₀, NamedTuple(), 0)
```
Comment (lukem12345) [inline test/simulation.jl:145]: Test `@allocations` as well

File: src/simulation.jl [hunk 2]
```diff
@@ -688,7 +684,7 @@ state variables and literals in the Decapode.
 Optional keyword arguments are `dimension`, which is the dimension of the problem and defaults to 2D, `stateeltype`, which is the element type of the state forms and
 defaults to Float64 and `code_target`, which is the intended architecture target for the generated code, defaulting to regular CPU compatible code.
 """
-function gensim(user_d::SummationDecapode, input_vars::Vector{Symbol}; dimension::Int=2, stateeltype::DataType = Float64, code_target::AbstractGenerationTarget = CPUTarget())
+function gensim(user_d::SummationDecapode, input_vars::Vector{Symbol}; dimension::Int=2, stateeltype::DataType = Float64, code_target::AbstractGenerationTarget = CPUTarget(), can_prealloc::Bool = true)
```
Comment (lukem12345) [inline src/simulation.jl:687]: The docstring needs to be updated to document this flag.

File: test/simulation.jl [hunk 3]
```diff
@@ -412,17 +428,32 @@ end
...
 
-  sim = eval(gensim(contract_with_op2))
-  f = sim(earth, default_dec_generate)
-  A = 3 * ones(nv(earth))
-  E_dec = ones(nv(earth))
-  u = ComponentArray(A=A, E=E_dec)
-  du = ComponentArray(A=zeros(ntriangles(earth)), E=zeros(nv(earth)))
-  constants_and_parameters = ()
-  f(du, u, constants_and_parameters, 0)
+  let sim = eval(gensim(contract_with_op2))
```
Comment (lukem12345) [inline test/simulation.jl:431]: Find some way eliminating duplicated code. Loop over `[true,false]` for the value of the `can_prealloc` flag.

File: test/simulation.jl [hunk 4]
```diff
@@ -509,19 +540,34 @@ end
     F == A ∧ (C ∧ B)
   end
 
-  sim = eval(gensim(wedges01))
+  let sim = eval(gensim(wedges01))
```
Comment (lukem12345) [inline test/simulation.jl:543]: It seems the only difference here is the value of  `can_prealloc`; loop here as well.

File: examples/chemistry/brusselator_implicit.jl [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  α = 0.001,
+  F = t -> t ≥ 1.1 ? F₁ : F₂)
+
+sim = evalsim(Brusselator, can_prealloc=false)
+fₘ = sim(sd, nothing, DiagonalHodge())
+
+u₀ = ComponentArray(U=U, V=V)
+
+du₀ = copy(u₀)
+jac_sparsity = Symbolics.jacobian_sparsity((du, u) -> fₘ(du, u, constants_and_parameters, 0.0), du₀, u₀)
```
Comment (lukem12345) [inline examples/chemistry/brusselator_implicit.jl:67]: Can we copy these 10 or so lines into the existing `brusselator.jl` file instead

File: examples/chemistry/brusselator_implicit.jl [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+f = ODEFunction(fₘ; jac_prototype = float.(jac_sparsity))
+# tₑ = 11.5
+tₑ = 20
+@info("Solving")
+prob = ODEProblem(f, u₀, (0, tₑ), constants_and_parameters)
+soln = solve(prob, FBDF(linsolve = KLUFactorization()), progress=true, progress_steps=1, saveat=0.1);
```
Comment (lukem12345) [inline examples/chemistry/brusselator_implicit.jl:74]: Leave a comment if you expect this to take longer than other solvers and that this is just for demonstration purposes.

File: examples/chemistry/brusselator_implicit.jl (hunk 5 above)
Comment (GeorgeR227) [reply examples/chemistry/brusselator_implicit.jl:67]: I feel like all the Brusselator stuff we have is getting a bit bloated. I'm leaning towards collecting all these examples and turning them into a docs page after this PR is merged.

Comment (lukem12345) [reply examples/chemistry/brusselator_implicit.jl:67]: Ok. Move these lines to `brusselator.jl` so history is better preserved when we `git mv` it later.

File: examples/chemistry/brusselator_implicit.jl (hunk 6 above)
Comment (GeorgeR227) [reply examples/chemistry/brusselator_implicit.jl:74]: I've shrunk it down a bit and it only takes 22 seconds to solve.

Comment (lukem12345) [reply examples/chemistry/brusselator_implicit.jl:74]: What is the relative time to the other solver

File: examples/chemistry/brusselator_implicit.jl (hunk 5 above)
Comment (GeorgeR227) [reply examples/chemistry/brusselator_implicit.jl:67]: I should also note I changed the `F` function to match the one [here](https://docs.sciml.ai/DiffEqDocs/stable/tutorials/advanced_ode_example/). The one in `brusselator.jl` has a flipped time condition.

Comment (lukem12345) [reply examples/chemistry/brusselator_implicit.jl:67]: Another reason to have them in the same file so these differences are more apparent

File: src/simulation.jl (hunk 2 above)
Comment (GeorgeR227) [reply src/simulation.jl:689]: I've updated some other docstrings here that needed it.

Comment (lukem12345) [reply src/simulation.jl:689]: Why not here?

File: test/simulation.jl (hunk 1 above)
Comment (GeorgeR227) [reply test/simulation.jl:145]: Number of allocations for new is reasonable but I can't really understand the size of the allocations. It's less than I expect.

File: examples/chemistry/brusselator_implicit.jl (hunk 6 above)
Comment (GeorgeR227) [reply examples/chemistry/brusselator_implicit.jl]: Quite a bit longer, but I want to showcase how to use implicit methods so I've tagged this as demonstration code.

File: src/simulation.jl [hunk 7]
```diff
@@ -680,15 +677,16 @@ end
...
-    gensim(user_d::SummationDecapode, input_vars::Vector{Symbol}; dimension::Int=2, stateeltype::DataType = Float64, code_target::GenerationTarget = CPUTarget())
+    gensim(user_d::SummationDecapode, input_vars::Vector{Symbol}; dimension::Int=2, stateeltype::DataType = Float64, code_target::AbstractGenerationTarget = CPUTarget(), can_prealloc::Bool = true)
 
 Generates the entire code body for the simulation function. `user_d` is the user passed Decapodes which will be left unmodified and 'input_vars' is the collection of
 state variables and literals in the Decapode.
 
 Optional keyword arguments are `dimension`, which is the dimension of the problem and defaults to 2D, `stateeltype`, which is the element type of the state forms and
-defaults to Float64 and `code_target`, which is the intended architecture target for the generated code, defaulting to regular CPU compatible code.
+defaults to Float64, `code_target`, which is the intended architecture target for the generated code, defaulting to regular CPU compatible code, and 'can_prealloc' which
+is set to `true` by default and determines if intermediate results can be preallocated.
```
Comment (lukem12345) [inline src/simulation.jl:687]: Is setting this flag necessary for using implicit solvers? And if so can that be documented here?

Comment (GeorgeR227) [reply src/simulation.jl:687]: No, but if you want to find a sparse Jacobian pattern, plus some other more involved implicit solver things I forget exactly what, then you do need this on.

Comment (GeorgeR227) [reply src/simulation.jl:687]: But this is more so a "safe" mode, where there's less optimization stuff going on that could impact solving, such as needing to find a sparse Jacobian.

Comment (lukem12345) [reply src/simulation.jl:687]: Can you document that either in this docstring or in a blurb in the Docs page

Comment (GeorgeR227) [reply src/simulation.jl:687]: Yeah, I can do that when I make a docs page on Brusselator.

Comment (lukem12345) [reply src/simulation.jl:687]: Ok. For the time being can you put something in this doctoring then.

Comment (GeorgeR227) [reply src/simulation.jl]: Done and I've also cleaned up the whole docstring since this'll be important to the user.

Comment (lukem12345) [reply src/simulation.jl]: I mean can you mention in the docstring the "why" of this flag (Get JacSparsity to work etc.). The "what" is excellently answered.

**Comment by jpfairbanks** [comment]
This is a great feature! The name can_prealloc is a little weird I would like preallocation or a standardized name from the SciML ecosystem if there is one.

**Comment by GeorgeR227** [comment]
I couldn't find anything exactly from SciML but they seem to favor using whole words for keywords. I think `preallocate` would be good because it's a verb and gets across the idea that you tell gensim to either do it (true) or not (false).

**Review by lukem12345 — APPROVED** [review]

**Pull request merged**